tutorials/cloud-functions-async/index.md

```diff
@@ -0,0 +1,150 @@
+
+This is a basic pattern which may have some use-case specific details.
+
+## Serverless Stateful Jobs
+
+This tutorial uses several managed services to implement this pattern including [PubSub](https://cloud.google.com/pubsub/) and [Firestore](https://cloud.google.com/firestore/).
+
+![](image/arch.png)
+
+You can implement job state storage or a work queue with alternatives like Cloud SQL and Cloud Tasks. This tutorial uses the support of [golang](https://golang.org/) in Cloud Functions, but any supported language can be used to fulfill the pattern.
+
+### Job resource definition
+
+A simple custom job definition is defined:
+
+```
+type JobState int
+
+const (
+	Created JobState = iota
+	Running
+	Completed
+	Failed
+)
+
+type Job struct {
+	ID         string    `json:"id"`
+	CreateTime time.Time `json:"created-time"`
+	DoneTime   time.Time `json:"done-time" firestore:"DoneTime,omitempty"`
+	Done       bool
+	Result     string
+	State      JobState
+	Task       map[string]interface{} `json:"-"`
+	// can add a requester, source IP etc if needed
+}
+```
+
+When a job request is received, it is given an ID and the details of the work are included in a task field. This task is sent into the work queue as PubSub payload, with the Job ID as an [PubSub attribute](https://cloud.google.com/pubsub/docs/publisher#custom-attributes). When a worker picks up the task, it moves the job from `Created` -> `Running`. When the task is complete, it moves from `Running` to `Completed` (success) or `Failed` and writes any result back into the Firestore document.
+
+## Set Up
+
+1.  Create a project in the [Google Cloud Platform Console][console].
+1.  Enable billing for your project.
+1.  Use [Cloud Shell][shell] or Install the [Google Cloud SDK][sdk].
+
+[console]: https://console.cloud.google.com/
+[shell]: https://cloud.google.com/shell/
+[sdk]: https://cloud.google.com/sdk/
+
+Commands in this tutorial assume you are running from the tutorial folder:
+
+```
+git clone https://github.com/GoogleCloudPlatform/community.git
+cd community/tutorials/cloud-functions-async
+```
+
+### Create pubsub resources
+```
+gcloud pubsub topics create jobs
+gcloud pubsub subscriptions create worker-tasks --topic jobs
+```
+
+### Deploy Function
+
+`gcloud functions deploy Jobs --runtime go111 --trigger-http`
+
+## Create Jobs and check their state
+
+Once the function is deployed, we can use both POST and GET methods to interact with long running jobs.
+
+### Get URL
+
+Get the URL of the deployed function into an environment variable:
+
+```
+export URL=$(gcloud functions describe Jobs --format='value(httpsTrigger.url)')
+```
+
+### Create a Job
+
+Call the function based API to create a Job with a simple task. Our task has one field which is `worktime` which determines how many seconds the task takes to complete.
```

Comment:
function based -> function-based

Our task has one field which is `worktime` which determines how many seconds the task takes to complete. - > Our task has one field, `worktime`, which determines how many seconds the task takes to complete.

Reply:
done